## C - Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?


## Steps

Import all the libraries

In [70]:
import pandas as pd
import numpy as np
import sklearn.metrics as metrics
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense

import warnings
warnings.simplefilter('ignore', FutureWarning)

## Reading the data

In [71]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [72]:
# Getting the features and targets
features = concrete_data.drop(columns='Strength', axis=1)
Y = concrete_data['Strength']

features = (features - features.mean()) / features.std()

n_cols = features.shape[1] # number of predictors

x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)


#### 2. Train the model on the training data using 50 epochs.

In [73]:
def regression_model():
    # create model
    model = Sequential()
    #One hidden layer of 10 nodes, and a ReLU activation function 
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) 
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model = regression_model()
model

#### 2. Train the model on the training data using 50 epochs.

In [74]:
model.fit(x_train, y_train, epochs=100, verbose=2)

Epoch 1/100
 - 0s - loss: 1606.1620
Epoch 2/100
 - 0s - loss: 1589.5026
Epoch 3/100
 - 0s - loss: 1573.5584
Epoch 4/100
 - 0s - loss: 1558.2368
Epoch 5/100
 - 0s - loss: 1543.5281
Epoch 6/100
 - 0s - loss: 1528.4974
Epoch 7/100
 - 0s - loss: 1514.2018
Epoch 8/100
 - 0s - loss: 1499.5155
Epoch 9/100
 - 0s - loss: 1485.0476
Epoch 10/100
 - 0s - loss: 1470.6934
Epoch 11/100
 - 0s - loss: 1455.7197
Epoch 12/100
 - 0s - loss: 1440.9558
Epoch 13/100
 - 0s - loss: 1425.7193
Epoch 14/100
 - 0s - loss: 1410.3608
Epoch 15/100
 - 0s - loss: 1394.2230
Epoch 16/100
 - 0s - loss: 1377.8621
Epoch 17/100
 - 0s - loss: 1360.9694
Epoch 18/100
 - 0s - loss: 1343.1541
Epoch 19/100
 - 0s - loss: 1324.3847
Epoch 20/100
 - 0s - loss: 1304.7746
Epoch 21/100
 - 0s - loss: 1283.9550
Epoch 22/100
 - 0s - loss: 1262.1492
Epoch 23/100
 - 0s - loss: 1239.6290
Epoch 24/100
 - 0s - loss: 1216.5206
Epoch 25/100
 - 0s - loss: 1191.8499
Epoch 26/100
 - 0s - loss: 1167.1399
Epoch 27/100
 - 0s - loss: 1141.7237
Epoch 28/1

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [75]:
predictions = model.predict(x_test)
MAE =  metrics.mean_absolute_error(y_test, predictions)
print(MAE)

10.273411053160634


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [76]:
MAEs = []
for n in range(50): 
    x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)
    model.fit(x_train, y_train, epochs=50, verbose=0)
    predictions = model.predict(x_test)
    MAEs.append(metrics.mean_absolute_error(y_test, predictions))
    
print(MAEs)    

[8.782706684248346, 7.481691717252761, 7.201178205669119, 6.725530685251972, 6.466754623956279, 5.930085263545458, 5.514442297926227, 4.675257983037954, 4.995790659956947, 4.52508355396851, 4.8306659787372475, 4.763102711489286, 4.633209350919261, 4.662801941127839, 4.7063340701563074, 4.7734120590555635, 4.754899317423503, 4.741424530300893, 4.540023520015977, 4.232107414072771, 4.564791384576594, 4.732657494406098, 4.339123276213613, 4.525910980817184, 4.761543003687195, 4.537335615682756, 4.760456997183149, 4.554061974213734, 4.6413504888787624, 4.840275195161888, 4.519396685195972, 4.35927735325588, 4.552158611198846, 4.550830312870853, 4.440340799097104, 4.582664109140539, 4.448828945715451, 4.603360410153288, 4.531434761714009, 4.659254634202877, 4.513832569616126, 4.737096675946874, 4.79272999328317, 4.646631975389993, 4.524804813422045, 4.525849603634436, 4.593624059930203, 4.7721074971566315, 4.625599689730548, 4.40292961676144]


#### 5. Report the mean and the standard deviation of the mean squared errors.

In [77]:
np_array = np.array(MAEs)
print(f"Mean of mean squared errors:{np.mean(np_array)} ")
print(f"Standard deviation of mean squared errors:{np.std(np_array)} ")

Mean of mean squared errors:4.931535241926989 
Standard deviation of mean squared errors:0.8790470154158412 


### How does the mean of the mean squared errors compare to that from Step B?


Step A:

- Mean of mean squared errors:5.57246862771433 
- Standard deviation of mean squared errors:0.7217945033478287 

Step B:
 - Mean of mean squared errors:5.718177185425865 
 - Standard deviation of mean squared errors:1.2205812457213703 
 
Step C
 - Mean of mean squared errors:4.931535241926989 
 - Standard deviation of mean squared errors:0.8790470154158412 

Step C reduced a little the mean of mean squared errors and standard deviation if compared with B